x-pack/plugins/actions/server/lib/get_execution_log_aggregation.ts

```diff
@@ -0,0 +1,454 @@
+  buckets: IExecutionUuidKpiAggBucket[];
+}
+
+interface ExcludeExecuteStartAggResult extends estypes.AggregationsAggregateBase {
+  executionUuid: ExecutionUuidAggResult;
+  executionUuidCardinality: {
+    executionUuidCardinality: estypes.AggregationsCardinalityAggregate;
+  };
+}
+
+interface ExcludeExecuteStartKpiAggResult extends estypes.AggregationsAggregateBase {
+  executionUuid: ExecutionUuidKPIAggResult;
+}
+
+export interface IExecutionLogAggOptions {
+  filter?: string | KueryNode;
+  page: number;
+  perPage: number;
+  sort: estypes.Sort;
+}
+
+const ExecutionLogSortFields: Record<string, string> = {
+  timestamp: 'actionExecution>executeStartTime',
+  execution_duration: 'actionExecution>executionDuration',
+  total_search_duration: 'actionExecution>totalSearchDuration',
+  schedule_delay: 'actionExecution>scheduleDelay',
+};
+
+export const getExecutionKPIAggregation = (filter?: IExecutionLogAggOptions['filter']) => {
+  const dslFilterQuery: estypes.QueryDslBoolQuery['filter'] = buildDslFilterQuery(filter);
+
+  return {
+    excludeExecuteStart: {
+      filter: {
+        bool: {
+          must_not: [
+            {
+              term: {
+                'event.action': 'execute-start',
+              },
+            },
+          ],
+        },
+      },
+      aggs: {
+        executionUuid: {
+          // Bucket by execution UUID
+          terms: {
+            field: EXECUTION_UUID_FIELD,
+            size: DEFAULT_MAX_KPI_BUCKETS_LIMIT,
+            order: formatSortForTermSort([{ timestamp: { order: 'desc' } }]),
+          },
+          aggs: {
+            executionUuidSorted: {
+              bucket_sort: {
+                from: 0,
+                size: DEFAULT_MAX_KPI_BUCKETS_LIMIT,
+                gap_policy: 'insert_zeros' as estypes.AggregationsGapPolicy,
+              },
+            },
+            actionExecution: {
+              filter: {
+                bool: {
+                  ...(dslFilterQuery ? { filter: dslFilterQuery } : {}),
+                  must: [getProviderAndActionFilter('actions', 'execute')],
+                },
+              },
+              aggs: {
+                executeStartTime: {
+                  min: {
+                    field: START_FIELD,
+                  },
+                },
+                actionExecutionOutcomes: {
+                  terms: {
+                    size: 3,
+                    field: OUTCOME_FIELD,
+                  },
+                },
+              },
+            },
+            minExecutionUuidBucket: {
+              bucket_selector: {
+                buckets_path: {
+                  count: 'actionExecution._count',
+                },
+                script: {
+                  source: 'params.count > 0',
+                },
+              },
+            },
+          },
+        },
+      },
+    },
+  };
+};
+
+export function getExecutionLogAggregation({
```

Comment:
I think the query/agg for the action exec log can be a lot simpler than the one we need for the rule exec log because we are only looking at `provider: action` docs from the event log, so we don't need to be grouping by rule execution UUID and scanning across many documents. The only thing we have to reconcile is that there is a separate document for when the execution times out. We can discuss this offline if you want.

Something that would be cool for a followup issue is being able to keep track of the number of retries an action has and only showing one entry in the exec log with the number of retries.

Reply:
Resolved in this commit efc5a506870cb13ad58e467d52ea5f6bf2812af9